File: justonesandzeros/Postings/_Posted/2011/Four Eyes/FourEyes.vb
Imports Autodesk.Connectivity.WebServices
Imports Autodesk.Connectivity.WebServicesTools
Imports Autodesk.Connectivity.Extensibility.Framework

<Assembly: ApiVersion("4.0")> 
<Assembly: ExtensionId("EF55A79B-90DC-47AF-8894-332D536E4164")> 

Public Class FourEyesChecker
    Implements IWebServiceExtension

    ' implement IWebServiceExtension
    Public Sub OnLoad() Implements Autodesk.Connectivity.WebServices.IWebServiceExtension.OnLoad
        AddHandler DocumentServiceExtensions.UpdateFileLifecycleStateEvents.GetRestrictions, _
            AddressOf UpdateFileLifecycleStateEvents_GetRestrictions
    End Sub

    ' event handler
    Private Sub UpdateFileLifecycleStateEvents_GetRestrictions(ByVal sender As Object, ByVal e As UpdateFileLifeCycleStateCommandEventArgs)
        Try
            Dim service As IWebService = TryCast(sender, IWebService)
            If service Is Nothing Then
                Return
            End If

            Dim cred As New WebServiceCredentials_bugfix(service)
            Using mgr As New WebServiceManager(cred)
                Dim currentUserId As Long = mgr.SecurityService.SecurityHeader.UserId

                Dim defs As LfCycDef() = mgr.DocumentServiceExtensions.GetAllLifeCycleDefinitions()
                Dim releaseProcess As LfCycDef = defs.FirstOrDefault(Function(n) n.SysName = "Flexible Release Process")

                Dim reviewState As LfCycState = releaseProcess.StateArray.FirstOrDefault(Function(n) n.DispName = "For Review")

                Dim releaseState As LfCycState = releaseProcess.StateArray.FirstOrDefault(Function(n) n.DispName = "Released")

                Dim fileCollection As FileArray() = mgr.DocumentService.GetFilesByMasterIds(e.FileMasterIds)

                For i As Integer = 0 To fileCollection.Length - 1
                    CheckFile(fileCollection(i).Files, e.ToStateIds(i), currentUserId, reviewState, releaseState, e)
                Next
            End Using
        Catch
        End Try
    End Sub

    ' checks for a four eyes violation for a given file history
    Private Sub CheckFile(ByVal files As File(), ByVal toStateId As Long, ByVal currentUserId As Long, ByVal reviewState As LfCycState, ByVal releaseState As LfCycState, ByVal eventArgs As WebServiceCommandEventArgs)
        ' if we are not moving to released, don't event bother with the check
        If toStateId <> releaseState.Id Then
            Return
        End If

        Dim maxFile As File = files.First(Function(n) n.MaxCkInVerNum = n.VerNum)
        If maxFile.FileRev Is Nothing Then
            Return
        End If

        ' gather all the files in the revision and arrange them by version
        Dim filesInRev As IEnumerable(Of File) = From n In files Where n.FileRev.RevId = maxFile.FileRev.RevId Order By n.VerNum

        Dim filesArray As File() = filesInRev.ToArray()

        Dim reviewUserId As Long = -1
        For i As Integer = 1 To filesArray.Length - 1
            Dim f1 As File = filesArray(i - 1)
            Dim f2 As File = filesArray(i)

            ' compare two concecutive file versions to determine 
            ' where a state changed happened
            If f1.FileLfCyc IsNot Nothing AndAlso f2.FileLfCyc IsNot Nothing AndAlso f1.FileLfCyc.LfCycStateName <> f2.FileLfCyc.LfCycStateName AndAlso f2.VerNum - f1.VerNum = 1 Then
                ' f2 is a version where the state changed
                If f2.FileLfCyc.LfCycStateName = reviewState.DispName Then
                    reviewUserId = f2.CreateUserId
                End If
            End If
        Next

        If reviewUserId > 0 AndAlso currentUserId = reviewUserId Then
            ' the same person reviewed the file in an earlier version
            eventArgs.AddRestriction(New ExtensionRestriction(maxFile.Name, "File cannot be reviewed and released by the same person"))
        End If
    End Sub

End Class

File: justonesandzeros/Postings/_Posted/2011/Four Eyes/FourEyes.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Autodesk.Connectivity.WebServices;
using Autodesk.Connectivity.WebServicesTools;
using Autodesk.Connectivity.Extensibility.Framework;

[assembly: ApiVersion("4.0")]
[assembly: ExtensionId("CAEA1F8E-C1F2-4C63-84D0-7E62CDCE2408")]

namespace FourEyes
{
    public class FourEyesChecker : IWebServiceExtension
    {

        // implement IWebServiceExtension
        public void OnLoad()
        {
            DocumentServiceExtensions.UpdateFileLifecycleStateEvents.GetRestrictions += 
                new EventHandler<UpdateFileLifeCycleStateCommandEventArgs>(
                    UpdateFileLifecycleStateEvents_GetRestrictions);
        }

        // event handler
        void UpdateFileLifecycleStateEvents_GetRestrictions(
            object sender, UpdateFileLifeCycleStateCommandEventArgs e)
        {
            try
            {
                IWebService service = sender as IWebService;
                if (service == null)
                    return;

                WebServiceCredentials_bugfix cred = new WebServiceCredentials_bugfix(service);
                using (WebServiceManager mgr = new WebServiceManager(cred))
                {
                    long currentUserId = mgr.SecurityService.SecurityHeader.UserId;

                    LfCycDef [] defs = 
                        mgr.DocumentServiceExtensions.GetAllLifeCycleDefinitions();
                    LfCycDef releaseProcess = defs.FirstOrDefault(
                        n => n.SysName == "Flexible Release Process");

                    LfCycState reviewState = 
                        releaseProcess.StateArray.FirstOrDefault(
                        n => n.DispName == "For Review");

                    LfCycState releaseState = 
                        releaseProcess.StateArray.FirstOrDefault(
                        n => n.DispName == "Released");

                    FileArray [] fileCollection = 
                        mgr.DocumentService.GetFilesByMasterIds(e.FileMasterIds);

                    for (int i = 0; i < fileCollection.Length; i++)
                    {
                        CheckFile(fileCollection[i].Files, e.ToStateIds[i], 
                            currentUserId, reviewState, releaseState, e);
                    }
                }
            }
            catch { }
        }

        // checks for a four eyes violation for a given file history
        private void CheckFile(File[] files, long toStateId, 
            long currentUserId, LfCycState reviewState, 
            LfCycState releaseState, WebServiceCommandEventArgs eventArgs)
        {
            // if we are not moving to released, don't event bother with the check
            if (toStateId != releaseState.Id)
                return;

            File maxFile = files.First(n => n.MaxCkInVerNum == n.VerNum);
            if (maxFile.FileRev == null)
                return;

            // gather all the files in the revision and arrange them by version
            IEnumerable<File> filesInRev = 
                from n in files
                where n.FileRev.RevId == maxFile.FileRev.RevId
                orderby n.VerNum
                select n;

            File [] filesArray = filesInRev.ToArray();

            long reviewUserId = -1;
            for (int i = 1; i < filesArray.Length; i++)
            {
                File f1 = filesArray[i-1];
                File f2 = filesArray[i];

                // compare two concecutive file versions to determine 
                // where a state changed happened
                if (f1.FileLfCyc != null && f2.FileLfCyc != null &&
                    f1.FileLfCyc.LfCycStateName != f2.FileLfCyc.LfCycStateName &&
                    f2.VerNum - f1.VerNum == 1)
                {
                    // f2 is a version where the state changed
                    if (f2.FileLfCyc.LfCycStateName == reviewState.DispName)
                        reviewUserId = f2.CreateUserId;
                }
            }

            if (reviewUserId > 0 && currentUserId == reviewUserId)
            {
                // the same person reviewed the file in an earlier version
                eventArgs.AddRestriction(
                    new ExtensionRestriction(maxFile.Name, 
                    "File cannot be reviewed and released by the same person"));
            }
        }

    }
}
